lib/segment/src/segment_constructor/segment_constructor_base.rs

```diff
@@ -168,3 +191,61 @@ pub fn build_segment(path: &Path, config: &SegmentConfig) -> OperationResult<Seg
 
     Ok(segment)
 }
+
+fn load_segment_state(contents: &str, path: &Path) -> OperationResult<SegmentState> {
+    #[derive(Debug, Deserialize, Serialize, JsonSchema, Clone)]
+    #[serde(rename_all = "snake_case")]
+    pub struct ObsoleteSegmentState {
+        pub version: SeqNumberType,
+        pub config: ObsoleteSegmentConfig,
+    }
+
+    #[derive(Debug, Deserialize, Serialize, JsonSchema, Clone)]
+    #[serde(rename_all = "snake_case")]
+    pub struct ObsoleteSegmentConfig {
+        /// Size of a vectors used
+        pub vector_size: usize,
+        /// Type of distance function used for measuring distance between vectors
+        pub distance: Distance,
+        /// Type of index used for search
+        pub index: Indexes,
+        /// Type of vector storage
+        pub storage_type: StorageType,
+        /// Defines payload storage type
+        #[serde(default)]
+        pub payload_storage_type: PayloadStorageType,
+    }
+
+    let from_obsolete = |state: ObsoleteSegmentState| {
+        let vector_data = VectorDataConfig {
+            vector_size: state.config.vector_size,
+            distance: state.config.distance,
+        };
+        SegmentState {
+            version: state.version,
+            config: SegmentConfig {
+                vector_data: HashMap::from([(DEFAULT_VECTOR_NAME.to_owned(), vector_data)]),
+                index: state.config.index,
+                storage_type: state.config.storage_type,
+                payload_storage_type: state.config.payload_storage_type,
+            },
+        }
+    };
+
+    let segment_state = serde_json::from_str::<SegmentState>(contents).map_err(|err| {
+        OperationError::service_error(&format!(
+            "Failed to read segment {}. Error: {}",
+            path.to_str().unwrap(),
+            err
+        ))
+    });
+    if segment_state.is_err() {
+        // try to load obsolete format
+        match serde_json::from_str::<ObsoleteSegmentState>(contents) {
+            Ok(obsolete_segment_state) => Ok(from_obsolete(obsolete_segment_state)),
```

Comment:
We do not save a new version of config anywhere.

Reply:
Also, relying on parsing failure might be dangerous. Theoretically it is possible, that the parser won't break, but will return malformed result (e.g. if we would use default/optional fields). Maybe it might sense to have one function to load deprecated config and one to load new one and switch between those functions based on the segment crate version. + also always save migrated config version

Reply:
fixed